pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestMethodShouldHaveIsTestAnnotationRule.java

```diff
@@ -0,0 +1,58 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.rule.bestpractices;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+
+import net.sourceforge.pmd.lang.apex.ast.ASTMethod;
+import net.sourceforge.pmd.lang.apex.ast.ASTMethodCallExpression;
+import net.sourceforge.pmd.lang.apex.ast.ApexNode;
+import net.sourceforge.pmd.lang.apex.rule.AbstractApexUnitTestRule;
+
+public class ApexUnitTestMethodShouldHaveIsTestAnnotationRule extends AbstractApexUnitTestRule {
+    private static final Set<String> ASSERT_METHODS = new HashSet<>();
+    private static final String ASSERT = "system.assert";
+    private static final String ASSERT_EQUALS = "system.assertequals";
+    private static final String ASSERT_NOT_EQUALS = "system.assertnotequals";
+
+    static {
+        ASSERT_METHODS.add(ASSERT);
+        ASSERT_METHODS.add(ASSERT_EQUALS);
+        ASSERT_METHODS.add(ASSERT_NOT_EQUALS);
+    }
+
+    public ApexUnitTestMethodShouldHaveIsTestAnnotationRule() {
+        addRuleChainVisit(ASTMethod.class);
```

Comment:
As mentioned above: The base class `AbstractApexUnitTest` tries to apply this rule only for test classes. But adding a rule chain visit only for ASTMethod, makes this rule apply for any class.

Could this be a problem for this rule? It would report any method, that has a assert statement, regardless whether the method is contained in a test class or not...

If this is a problem, then we simply can't use the rule chain for this rule.

Are assert statements used outside of unit tests? If so, we should add a test case, where the class is a "normal" class (not a IsTest class) with a method, that contains a assert statement. This test case should not trigger the rule, since it is not a test class, and the method is therefore no test method.

Reply:
AFAIK, It is not a good practice to use assert statements in non test classes. I will remove the class `addRuleChainVisit` from both of the rules.